### PR TITLE
NuRandomService engine for geant4 seed generator

### PR DESCRIPTION
**Pull request by drivera-fnal**

### CODE REVIEW

**Comment by lgarren** [comment]
trigger build

**Comment by yangtj207** [comment]
I made a feature branch feature/tjyang_largeantseed in dunetpc to work with this PR.

File: larg4/Core/larg4Main_module.cc [hunk 1]
```diff
@@ -148,70 +150,67 @@ larg4::larg4Main::larg4Main(fhicl::ParameterSet const & p)
...
+  // ((artg4tk::EventActionBase*)&*pla) -> callArtProduces(this);
+  // ((artg4tk::TrackingActionBase*)&*pla) -> callArtProduces(this);
+
+  // -- Check for invalid seed value
+  if (seed_ > 900000000) {
+    //mf::LogError("SeedCheck") << "Bad seed provided, max seed value is 9E8.
+    throw cet::exception("largeant:BadSeedValue")
+          << "The provided largeant seed value: " << seed_
+          << " is invalid! Maximum seed value is 9E8.";
   }
```
Comment (knoepfel) [inline larg4/Core/larg4Main_module.cc:210]: This check is not necessary--it is already done by `art` when `createEngine` is called.

File: larg4/Core/larg4Main_module.cc [hunk 2]
```diff
@@ -148,70 +150,67 @@ larg4::larg4Main::larg4Main(fhicl::ParameterSet const & p)
...
-  createEngine( seed_, "G4Engine");
+  // Set up the random number engine.
+  // -- D.R.: Use the NuRandomService engine for additional control over the seed generation policy
+  (void)art::ServiceHandle<rndm::NuRandomService>()->createEngine(*this,"G4Engine",p,"seed");
```
Comment (knoepfel) [inline larg4/Core/larg4Main_module.cc:213]: Is there a reason you cannot simply call:
```.cpp
createEngine(p.get<long>("seed", 0), "G4Engine");
```
or some other default seed value.  If that function call is sufficient, then that's a lot simpler than introducing a dependency on `NuRandomService`, which has known multi-threading problems.

File: larg4/Core/larg4Main_module.cc (hunk 1 above)
Comment (drivera-fnal) [reply larg4/Core/larg4Main_module.cc:210]: There was no error thrown when I set the seed to 1E9 in the fhicl file. 

```
%MSG-i NuRandomService:  larg4Main:largeant@Construction  24-Mar-2020 20:07:36 CDT ModuleConstruction
Seeding G4Engine engine "largeant:" with seed 1000000000.
   ...
Summary of seeds computed by the NuRandomService
Random policy: 'perEvent'
  algorithm version: EventTimestamp_v1
  Configured value          Last value   ModuleLabel.InstanceName
         1000000000              (same)   largeant [overridden]
```

Comment (knoepfel) [reply larg4/Core/larg4Main_module.cc:210]: Which version of art were you using when running that job?

File: larg4/Core/larg4Main_module.cc (hunk 2 above)
Comment (drivera-fnal) [reply larg4/Core/larg4Main_module.cc:213]: The reason to use the NuRandomService `createEngine` was to have control over the engine policy. We are able to reproduce events in LegacyLArG4 following this [redmine wiki](https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Reproducing_jobs_using_the_same_random_number_sequences) with the `perEvent` policy.

We could use the native `createEngine` from the EDProducer base class; however, there are multiple wikis that say things like

> The preferred way to control random number streams is through NuRandomService

I wasn't aware that there were issues with multi-threading for the NuRandomService, however, and I know that @hanswenzel (and maybe others) have started looking into multi-threading.

File: larg4/Core/larg4Main_module.cc (hunk 1 above)
Comment (drivera-fnal) [reply larg4/Core/larg4Main_module.cc:210]: `v3_05_00` I'm building off of larsoft suite v08_47_00

File: larg4/Core/larg4Main_module.cc (hunk 2 above)
Comment (knoepfel) [reply larg4/Core/larg4Main_module.cc:213]: Thanks for the info, @drivera-fnal.  We should re-evaluate the guidance given in the redmine documentation...from a multi-threading perspective `NuRandom` is *not* the preferred way of handling random numbers.  

Let me talk with the SciSoft team regarding the best way forward here.  Adding `NuRandom` right now is suspect.  Do you mind opening up a support issue, where you mention what you require in terms of random numbers/reproducibility?

https://cdcvs.fnal.gov/redmine/projects/larsoft/issues/new

Thanks.

File: larg4/Core/larg4Main_module.cc (hunk 1 above)
Comment (knoepfel) [reply larg4/Core/larg4Main_module.cc:210]: Yikes!  I wouldn't have expected that.  The only way a value greater than 9E8 should be allowed is if you're using the `MixMaxRng` random number generator.  In that case, it's supported.  Are you by chance setting 
```
services.RandomNumberGenerator.defaultEngineKind: "MixMaxRng"
``` 
in your configuration file?

**Comment by hanswenzel** [comment]
Hi

some guidance would be nice. I am all for using MixMax and getting rid of all the yikes.
________________________________

File: larg4/Core/larg4Main_module.cc (hunk 2 above)
Comment (drivera-fnal) [reply larg4/Core/larg4Main_module.cc:213]: No problem @knoepfel 
I will fill out a new support issue with the details of our request.

File: larg4/Core/larg4Main_module.cc (hunk 1 above)
Comment (drivera-fnal) [reply larg4/Core/larg4Main_module.cc:210]: No, this parameter isn't set in my fcl file. I dumped it and this is all there is for the NuRandomService at least.
```
     NuRandomService: {
        endOfJobSummary: true
        policy: "perEvent"
     }
```

**Comment by drivera-fnal** [comment]
Here's the link to the support request on redmine: [Support #24226](https://cdcvs.fnal.gov/redmine/issues/24226#change-76365)

**Comment by knoepfel** [comment]
@drivera-fnal, thank you.  I will talk with Gianluca, the designer of `NuRandomService`.  The primary issue is that `NuRandomService` couples the setting of any registered engines with the generation of seeds.  I hope Gianluca and I can settle on a solution that naturally fits your needs.

**Comment by drivera-fnal** [comment]
Thanks for the update @knoepfel

**Comment by knoepfel** [comment]
trigger build

**Comment by lgarren** [comment]
trigger build

**Comment by lgarren** [comment]
@drivera-fnal @tomjunk @yangtj207 
There is a problem with one of the dunetpc CI tests
https://dbweb5.fnal.gov:8443/LarCI/app/ns:dune/build_detail/test_details?build_id=dune_ci/8626&platform=Linux%202.6.32-754.28.1.el6.x86_64&buildtype=slf6%20e19:py2:prof&phase=ci_tests&test=ci_g4_regression_test_protoDUNEsp

233: %MSG
234: %MSG-s ArtException:  larg4Main:largeant@Construction 13-Apr-2020 18:17:49 CDT  ModuleConstruction
235: cet::exception caught in art
236: ---- Configuration BEGIN
237:   A seed for the instance 'largeant' was requested, but there is no configuration at all for 'largeant' module label.
238: ---- Configuration END
239: %MSG
240: Art has completed and will exit with status 9.

**Comment by drivera-fnal** [comment]
Hi Lynn,

I can't seem to recreate this error.

This is using my PR branch (for larsoft suite v08_47_00) and running the
dunetpc v08_49_00 ci tests?

if so, should I merge the latest changes, or has that already been done
before running the tests?

Thanks,
David

**Comment by yangtj207** [comment]
Feature branch feature/tjyang_largeantseed in dunetpc is needed for the CI test.

**Comment by yangtj207** [comment]
Lynn, could you start a CI test with feature/tjyang_largeantseed in dunetpc? I do not know how to do it. Thanks.

**Comment by lgarren** [comment]
I've triggered the build manually.  For furture reference, the instructions are here:
https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Manually_trigger_a_CI_test

@drivera-fnal will any of the other experiments need feature branches?

**Comment by drivera-fnal** [comment]
@Lynn A Garren <garren@fnal.gov>  ProtoDUNE is the only experiment using
the refactored larg4, currently.

**Comment by lgarren** [comment]
Thanks David.

**Comment by lgarren** [comment]
The CI test is fine with the feature branch included. https://dbweb5.fnal.gov:8443/LarCI/app/ns:DUNE/view_builds/index?offset=0&builds=lar_ci/8277&builds=

@yangtj207 @tomjunk @aihimmel   We need level 2 approval before proceeding.

**Comment by yangtj207** [comment]
+1

**Comment by lgarren** [comment]
approve

**Comment by lgarren** [comment]
merged as part of the release

**Pull request closed without merging**